### DISCUSSION TITLE
Android performance improment

### DISCUSSION THREAD

**Discussion by byme8** [AvaloniaUI/Avalonia#8690] [General]
I have been playing around with Avalonia on android, and the performance is not looking great at all.
I create a simple test project. The full source code is here: https://github.com/byme8/AvaloniaPerformanceTest
The basic logic is pretty straightforward;
``` cs
private void Button_OnClick(object sender, RoutedEventArgs e)
{
    var panels = Enumerable
        .Range(0, 10)
        .Select(o =>
        {
            var stackPanel = new StackPanel()
            {
                Orientation = Orientation.Horizontal
            };

            stackPanel.Children.AddRange(Enumerable.Range(0, 10).Select(oo => new Label() { Content = "Test" + o + oo }));
            return stackPanel;
        });
    
    var container = new StackPanel()
    {
        Orientation = Orientation.Vertical,
    };
    
    container.Children.AddRange(panels);

    Presenter.Content = container;
}
```

It takes up to 2.5 seconds to show up. That is a lot. The Xamarin Forms/MAUI or HTML/CSS for the same test cases is much faster, close to 300ms.

So, I started to wonder what was wrong here. I connected the net6 event profiler to the android Avolonia app, It tells that the app spends ~1.3 seconds inside the `` avalonia.base!Avalonia.AvaloniaObject.VerifyAccess() `` method. If I am not mistaken, it is required to ensure that we do our stuff in the main thread. This stuff is necessary, but I guess there is huge room for improvement.

The simplest way is to create a sort of "safe zone" to ensure that we are inside the main thread and then stop checking it whatsoever.

Attaching profiler results in the speedscope format.

[avalonia_apptest_trace.speedscope.json.zip](https://github.com/AvaloniaUI/Avalonia/files/9275089/avalonia_apptest_trace.speedscope.json.zip)

**Comment by Gillibald** [comment]
If that is the case you could try to always return true here: https://github.com/AvaloniaUI/Avalonia/blob/master/src/Android/Avalonia.Android/AndroidThreadingInterface.cs#L79

If that results in better performance we need to change that implementation

**Reply by byme8** [reply]
I took the ControlCatalog app and added my test cases there. For some reason, compiled from the main branch, it works much faster and takes ~1 second to render. At the same time, avalonia.base!Avalonia.AvaloniaObject.VerifyAccess() takes 800ms. So, nothing new. The issue is there. When I replaced the line that you mentioned with `` public bool CurrentThreadIsLoopThread => true;``, the render finished within ~180ms.

So, it really helps a lot.

Attaching profiler results.
The avalonia-fake-main-thread-trace.speedscope.json contains faked main thread.
The avalonia-normal-main-thread-trace.speedscope.json current implementation.

[traces.zip](https://github.com/AvaloniaUI/Avalonia/files/9275695/traces.zip)

**Reply by byme8** [reply]
I decided to see how MAUI checks for the main thread. They do it a bit differently. Like that:
``` cs
public bool CurrentThreadIsLoopThread => Looper.MainLooper.IsCurrentThread;
```
Potentially it can save some time because there is only one interaction with the JVM. I updated Avalonia to use this approach and got better results.
The whole render happens within ~500-600ms and `` avalonia.base!Avalonia.AvaloniaObject.VerifyAccess()  `` takes 300-400ms. It is better. but still not ideal.

**Reply by maxkatz6** [reply]
@kekekeks proposed something like this:
```csharp
static Thread s_uiThread;
public bool CurrentThreadIsLoopThread 
{
    get
    {
         if(s_uiThread != null)
             return s_uiThread == Thread.CurrentThread;
         if(Looper.MainLooper.Thread.Equals(Java.Lang.Thread.CurrentThread()))
         {
              s_uiThread = Thread.CurrentThread;
              return true;
         }
         return false;
    }
}
```

It should save main thread locally and reuse it, avoiding JVM calls.

**Reply by maxkatz6** [reply]
We can replace `Equals` call with `Looper.MainLooper.IsCurrentThread` as well

**Reply by byme8** [reply]
Tested this fix. Now it takes ~180ms to render. Nice!

**Comment by danwalmsley** [comment]
Fixed: https://github.com/AvaloniaUI/Avalonia/pull/8691
